[WIP] QUARKUS-3363 - Offline startup for Hibernate ORM in Quarkus

sebersole opened the pull request:
https://issues.redhat.com/browse/QUARKUS-3363
Fixes #13522

Handle ORM DialectSpecificSettings

sebersole committed:
QUARKUS-3363 - Disable JDBC Metadata Defaults in quarkus

Handle Hibernate's `DialectSpecificSettings` as supported Quarkus settings

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/CockroachDialectConfig.java:
```diff
@@ -0,0 +1,25 @@
+ * @author Steve Ebersole
+ */
+@ConfigGroup
+public interface CockroachDialectConfig {
```
So... maybe that's what you were getting at on Zulip when asking what should be exposed? Which _DBs_ are supported on Quarkus?

Because the list is here: https://quarkus.io/guides/datasource#configure-a-single-datasource

```
DB2: db2
Derby: derby
H2: h2
MariaDB: mariadb
Microsoft SQL Server: mssql
MySQL: mysql
Oracle: oracle
PostgreSQL: postgresql, pgsql or pg
```

Cockroach, HANA and Sybase are not officially supported in Quarkus -- as you may have noticed when you looked at `DatabaseKindDialectBuildItem`. I think it would be better to leave those configurable through `unsupported-properties` ATM, otherwise this will cause confusion as to which DBs are supported and which aren't...

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:
```diff
@@ -310,13 +310,52 @@ interface HibernateOrmConfigPersistenceUnitDialect {
          *
          * E.g. `MyISAM` or `InnoDB` for MySQL.
          *
+         * @deprecated Use {@code mysql.}{@linkplain MySQLDialectConfig#storageEngine storage-engine} instead
+         *
          * @asciidoclet
```
I am not sure using Javadoc tags such as `@code` and `@linkplain` will work with `@asciidoclet`... Though it probably doesn't matter in this specific case, as deprecated configuration properties are not included in documentation.

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmConfigPersistenceUnit.java:
```diff
@@ -310,13 +310,52 @@ interface HibernateOrmConfigPersistenceUnitDialect {
+         */
+        HANADialectConfig hana();
+
+        /**
+         * Configuration specific to the Hibernate ORM {@linkplain org.hibernate.dialect.MySQLDialect}
+         */
+        MySQLDialectConfig mysql();
```
Note we have use dedicated configuration for MariaDB in Quarkus, e.g. there is a `db-kind` named `mariadb`, distinct from `mysql`. It would probably make sense to do the same here?

FWIW you *might* be able to use inheritance with config groups if you want to avoid duplicating code... I think Guillaume added that support in his recent changes to the config annotation processor. We'll need to double checked by having a look at the [resulting documentation](https://quarkus-pr-main-43396-preview.surge.sh/version/main/guides/hibernate-orm#hibernate-configuration-properties), and of course through tests.

sebersole replied:
I could just add `MySQLDialectConfig mariadb();`, no?

yrodiere replied:
True, for now this would work.

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -1174,24 +1186,103 @@ private static void collectDialectConfig(String persistenceUnitName,
+            Optional<String> dbKind,
+            Optional<String> dialect) {
+
+        // todo : do we want dbKind/dialect aware handling for all of these settings (i.e. isMySQLOrMariaDB)?
+        //      if so, we need to reorganize this code a bit
```
When we know for sure what the db-kind/dialect is, indeed it would probably be a good idea to at least warn users if they set dialect options that are not going to have any effect -- like we do for `storage-engine`. We can even consider throwing an exception, but will need to mention this in the migration guide.

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/MySQLDialectConfig.java:
```diff
@@ -0,0 +1,46 @@
+     * Specifies the bytes per character to use based on the database's configured
+     * <a href="https://dev.mysql.com/doc/refman/8.0/en/charset-charsets.html">charset</a>.
+     *
+     * @see org.hibernate.cfg.DialectSpecificSettings#MYSQL_BYTES_PER_CHARACTER
```
Note:

1. Such links won't appear in generated documentation.
2. Application users are unlikely to reach this javadoc directly, they'll most likely read the rendered version included in the Quarkus documentation -- and in IDE plugins.

That's why we tend to do things like this: https://github.com/quarkusio/quarkus/blob/eeaecac7573dced42fa73c067ec84602c75120e3/extensions/hibernate-search-orm-outbox-polling/runtime/src/main/java/io/quarkus/hibernate/search/orm/outboxpolling/runtime/HibernateSearchOutboxPollingBuildTimeConfigPersistenceUnit.java#L88-L90
Reminder that Asciidoc syntax requires `@asciidoctlet` in the Javadoc comment.

`hibernate-orm-docs-url` is defined [here](https://github.com/quarkusio/quarkus/blob/b97f08b1d8642f184095cb704bc08eb0d96a6c14/docs/src/main/asciidoc/_attributes.adoc#L54), in case you want to add links to Hibernate ORM Javadocs -- but since nowadays you have a list of documentation properties in the Hibernate ORM user guide, I guess it could be more practical to link directly there?

sebersole replied:
> Reminder that Asciidoc syntax requires @asciidoctlet in the Javadoc comment.

+1, sure just have not been using asciidoc formatting yet.  But clearly I will need to start :)

sebersole replied:
> but since nowadays you have a list of documentation properties in the Hibernate ORM user guide, I guess it could be more practical to link directly there?

I think that's a good idea, Currently we don't generate url anchors for each setting as we'd need to, but https://hibernate.atlassian.net/browse/HHH-18654

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/OracleDialectConfig.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Configuration specific to the Hibernate ORM {@linkplain org.hibernate.dialect.OracleDialect}
+ *
+ * @author Steve Ebersole
```
Fair warning: `@author` tags are "disallowed" in Quarkus, see here for a rationale: https://github.com/quarkusio/quarkus/blob/34d15fc919adccfda3f9280c4100bd0c052e3881/CONTRIBUTING.md#L190-L193

I'm not sure if they automatically get checked or not, but you'll probably want to get rid of them.

sebersole replied:
> Fair warning: @author tags are "disallowed" in Quarkus, see here for a rationale:

No problem.  My IDE just adds them

yrodiere reviewed (commented):
Thanks! I added a few comments below, we can discuss on Zulip if you want.

Regarding documentation, it would probably make sense to add a subsection under "Dialects > Supported databases"; see [here for the source](https://github.com/quarkusio/quarkus/blob/45c1c6e4f4e43f6a1ea7810a20099c07a1ef1438/docs/src/main/asciidoc/hibernate-orm.adoc?plain=1#L171), [here for the rendered version for this PR](https://quarkus-pr-main-43396-preview.surge.sh/version/main/guides/hibernate-orm#hibernate-dialect-supported-databases) (updates take a few minutes/hours after you push).
Note the subsection we'd add to the documentation does not need to list all configuration properties -- that's already done in [this automatically-generated list](https://quarkus-pr-main-43396-preview.surge.sh/version/main/guides/hibernate-orm#hibernate-configuration-properties). We would just need to mention that some dialects can be configured in more details to match the DB configuration, and link to the list.

Regarding tests, I think your only solution ATM would be to add something under `integration-tests/jpa-oracle`, and `integration-tests/jpa-mysql`, and so on.

~~You can take inspiration from `DefaultSchemaTest` for your test; I picked this as an example because it overrides some settings compared to other tests, which I think you will need to do. Relevant classes:~~

* ~~[`DefaultSchemaTest`](https://github.com/quarkusio/quarkus/blob/ca3f7d8c442cfcf3156d9bbed4b8a14b81156a44/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/defaultcatalogandschema/DefaultSchemaTest.java)~~
* ~~[`DefaultSchemaInGraalITCase`](https://github.com/quarkusio/quarkus/blob/ca3f7d8c442cfcf3156d9bbed4b8a14b81156a44/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/defaultcatalogandschema/DefaultSchemaInGraalITCase.java#L6) (same test extending the first, but run in native mode)~~
* ~~[main package with the entity and REST resource being tested](https://github.com/quarkusio/quarkus/tree/18d103499400bbeb7aafd296672fe7128e2f42cf/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/defaultcatalogandschema): yes it's a bit weird, but that's how we can test even a native image, where the binary cannot (reasonably) run JUnit by itself.~~

EDIT: Actually, since we're talking about build-time properties here, this won't work. I think your only option would be to add a second persistence unit to `application.properties`. You may need to have this new PU start offline if the settings you need to test won't work with the DB we test against -- but that's no big deal, as you mainly want to make sure the Dialect's state matches the settings you set.

Starting a PU offline will require an additional config property though, to explicitly ask Hibernate ORM to "start offline"... which was the point of https://github.com/quarkusio/quarkus/issues/13522, and the reason we're introducing all these dialect-specific settings.

sebersole committed:
QUARKUS-3363 - Disable JDBC Metadata Defaults in quarkus

Handle Hibernate's `DialectSpecificSettings` as supported Quarkus settings

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -1174,24 +1187,233 @@ private static void collectDialectConfig(String persistenceUnitName,
+    }
+
+    private static SupportedDatabaseKind determineDatabaseKind(String name) {
+        if ( name.contains( "db2" ) ) {
```
`DatabaseKind.isDB2(name)` seems more appropriate? Same for others.

sebersole replied:
No, because this is also called with the Dialect name.

yrodiere replied:
I think this will not work well for `db-kind` then, because some database kinds (e.g. `postgresql`) have more than one corresponding string (e.g. `pg`). You'll probably need two different classes.

For what it's worth, we already have code that infers the "db product name" (the Hibernate ORM one) from the explicit dialect here:

https://github.com/quarkusio/quarkus/blob/c09e69a2f7411326108465d2e50a94c489defbad/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L1137-L1141

So maybe you should base your code on this `dbProductName` instead?

sebersole replied:
> I think this will not work well for db-kind then, because some database kinds (e.g. postgresql) have more than one corresponding string (e.g. pg).

Look at the callers.  One works with a passed "database kind" and manually does the normalize which handles the case you were talking about (multiple names).  The other works with the Dialect name.

> So maybe you should base your code on this dbProductName instead?

Maybe, I'll take a look.

sebersole replied:
Currently this code handles (1) db-kind and (2) explicit Dialect name.

Are you saying that it is legal to specify a dbProductName but not db-kind?

sebersole replied:
So the "problem" with using dbProductName is that, unless Quarkus does this somewhere, there is no simple mapping of dbProductName to Dialect or SupportedDatabaseKind.

Assuming that it is legal to have just dbProductName without either db-kind or dialect, we'd need a method that handles mapping dbProductName to a SupportedDatabaseKind

sebersole replied:
Ok, I added handling based on database product name as well.

yrodiere replied:
Thanks. Then if tests pass, we're all set.

FYI, the only way the database product name can be missing is if someone has no assigned datasource (because it's dynamic, see database multi-tenancy). Even then, users will need to set a dialect explicitly, and as long as they use one of the "standard" dialects, the database product name will be inferred from that dialect.

sebersole committed:
QUARKUS-3363 - Offline startup for Hibernate ORM in Quarkus

Handle Hibernate's `DialectSpecificSettings` as supported Quarkus settings

- test for mariadb

sebersole commented:
Was the integration test I added run?  I'm not sure which Action check runs those integration tests.

yrodiere commented:
> Was the integration test I added run? I'm not sure which Action check runs those integration tests.

CI does not run on drafts, you need to mark as ready for review and add `[WIP]` to the title. I'll do it.

yrodiere commented:
> > Was the integration test I added run? I'm not sure which Action check runs those integration tests.
> 
> CI does not run on drafts, you need to mark as ready for review and add `[WIP]` to the title. I'll do it.

Though the recommended way of working if your code is not ready for review is to run GitHub Actions in your fork instead.

sebersole commented:
> Imports are not sorted in /home/runner/work/quarkus/quarkus/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java

Um, I ran the formatter before I pushed.  Even now, locally, I run the formatter and nothing changes.  

What do I need to do here?  I find it strange that the IJ format stuff does not handle this either?

yrodiere commented:
> > Imports are not sorted in /home/runner/work/quarkus/quarkus/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
> 
> Um, I ran the formatter before I pushed. Even now, locally, I run the formatter and nothing changes.
> 
> What do I need to do here? I find it strange that the IJ format stuff does not handle this either?

How did you format code?

The formatter does not handle import order, that's a different setting.

If you installed the Eclipse Formatter plugin, you need to point to the import order config explicitly, it's a different file.
If you format from the command line, you must make sure to call the right goal/plugin.

sebersole commented:
I followed the directions from CONTRIBUTING : 

```
Formatting

Open the Preferences window (or Settings depending on your edition), navigate to Plugins and install the [Adapter for Eclipse Code Formatter](https://plugins.jetbrains.com/plugin/6546-eclipse-code-formatter) from the Marketplace.

Restart your IDE, open the Preferences (or Settings) window again and navigate to Adapter for Eclipse Code Formatter section on the left pane.

Select Use Eclipse's Code Formatter, then change the Eclipse workspace/project folder or config file to point to the eclipse-format.xml file in the independent-projects/ide-config/src/main/resources directory. Make sure the Optimize Imports box is ticked. Then, select Import Order from file and make it point to the eclipse.importorder file in the independent-projects/ide-config/src/main/resources directory.
```

sebersole commented:
I "format" using `mvnw -f extensions/hibernate-orm/deployment net.revelc.code.formatter:formatter-maven-plugin:2.24.1:format` because that's what the actions suggested earlier.

yrodiere commented:
> I "format" using `mvnw -f extensions/hibernate-orm/deployment net.revelc.code.formatter:formatter-maven-plugin:2.24.1:format` because that's what the actions suggested earlier.

I suppose the formatting plugin failed and suggested it. Here it's another plugin failing.

To run formatting + import sorting:

https://github.com/quarkusio/quarkus/blob/4eee652ec49b4fbe8abfca25e3cee3f16ebd5b23/CONTRIBUTING.md?plain=1#L273

Or if you want to limit to one module:

```
./mvnw process-sources -f extensions/hibernate-orm
```

yrodiere commented:
This part should have ensured that imports get sorted when you run auto-formatting in your IDE:

> Then, select Import Order from file and make it point to the eclipse.importorder file in the independent-projects/ide-config/src/main/resources directory.

There something wrong if CTRL+ALT+O / CTRL+ALT+L still doesn't format code correctly...

yrodiere commented on integration-tests/jpa-mariadb/src/main/resources/application.properties:
```diff
@@ -4,3 +4,12 @@ quarkus.datasource.password=hibernate_orm_test
+
+quarkus.datasource."offline".db-kind=mariadb
+quarkus.datasource."offline".db-version=12.0
+quarkus.hibernate."offline".mariadb.bytesPerCharacter=1
+quarkus.hibernate."offline".mariadb.noBackslashEscapes=true
+quarkus.hibernate."offline".mariadb.storageEngine=MY_CUSTOM_ENGINE
```
```suggestion
quarkus.hibernate-orm."offline".mariadb.bytesPerCharacter=1
quarkus.hibernate-orm."offline".mariadb.noBackslashEscapes=true
quarkus.hibernate-orm."offline".mariadb.storageEngine=MY_CUSTOM_ENGINE
```

yrodiere reviewed (commented):
Hey @sebersole , sorry I left this aside... were you waiting for feedback from me?

From what I can see you finished implementing most setting, and were trying to run tests.

Though I suspect in order to run tests, we will need to introduce the config property that QUARKUS-3363 actually is about (the dialect config just happens to be a prerequisite): something that tells Hibernate ORM to start offline, i.e. that sets `hibernate.boot.allow_jdbc_metadata_access` to `false` (on runtime startup), and probably triggers exceptions if users requested automatic schema creation/update/validation.
